coderd/database/dbfake/dbfake.go

```diff
@@ -39,44 +41,83 @@ func Workspace(t testing.TB, db database.Store, seed database.Workspace) databas
 
 // WorkspaceWithAgent is a helper that generates a workspace with a single resource
 // that has an agent attached to it. The agent token is returned.
-func WorkspaceWithAgent(t testing.TB, db database.Store, seed database.Workspace) (database.Workspace, string) {
+func WorkspaceWithAgent(
+	t testing.TB, db database.Store, seed database.Workspace,
+	mutations ...func([]*sdkproto.Agent) []*sdkproto.Agent,
+) (
+	database.Workspace, string,
+) {
 	t.Helper()
 	authToken := uuid.NewString()
+	agents := []*sdkproto.Agent{{
+		Id: uuid.NewString(),
+		Auth: &sdkproto.Agent_Token{
+			Token: authToken,
+		},
+	}}
+	for _, m := range mutations {
+		agents = m(agents)
+	}
 	ws := Workspace(t, db, seed)
 	WorkspaceBuild(t, db, ws, database.WorkspaceBuild{}, &sdkproto.Resource{
-		Name: "example",
-		Type: "aws_instance",
-		Agents: []*sdkproto.Agent{{
-			Id: uuid.NewString(),
-			Auth: &sdkproto.Agent_Token{
-				Token: authToken,
-			},
-		}},
+		Name:   "example",
+		Type:   "aws_instance",
+		Agents: agents,
 	})
 	return ws, authToken
 }
 
-// WorkspaceBuild inserts a build and a successful job into the database.
-func WorkspaceBuild(t testing.TB, db database.Store, ws database.Workspace, seed database.WorkspaceBuild, resources ...*sdkproto.Resource) database.WorkspaceBuild {
-	t.Helper()
+type BuildBuilder struct {
+	t         testing.TB
+	db        database.Store
+	ps        pubsub.Pubsub
+	ws        database.Workspace
+	seed      database.WorkspaceBuild
+	resources []*sdkproto.Resource
+}
+
+func WorkspaceBuildBuilder(t testing.TB, db database.Store, ws database.Workspace) BuildBuilder {
+	return BuildBuilder{t: t, db: db, ws: ws}
+}
```

Comment:
Naming suggestion

`BuildBuilder` sounds confusing, and may suggest that it creates a generic build that includes workspace build.

Maybe `NewWorkspaceBuildBuilder(<required params>)` and keep `WorkspaceBuildBuilder` as struct? It could be also a separate package `dbbuilders` if `dbfake` starts growing.

Reply:
I'd eventually like to make all of dbfake based on the builder pattern, and drop `Builder` and `New` from the public function names, but in the interim, we can go with the more-wordy `NewXxxxxxBuilder()` as the convention.